vivisect/impemu/platarch/i386.py

```diff
@@ -12,4 +17,52 @@ class i386WorkspaceEmulator(v_i_emulator.WorkspaceEmulator, e_i386.IntelEmulator
     def __init__(self, vw, logwrite=False, logread=False):
         e_i386.IntelEmulator.__init__(self)
         v_i_emulator.WorkspaceEmulator.__init__(self, vw, logwrite=logwrite, logread=logread)
-        self.setEmuOpt('i386:reponce',True)
+        self.setEmuOpt('i386:reponce', True)
+
+    def getRegister(self, index):
+        rval = value = e_i386.IntelEmulator.getRegister(self, index)
+
+        if self.op is None:
+            return value
+
+        ridx = index
+        if self.isMetaRegister(index):
+            ridx = index & RMETA_NMASK
+            # use the value of the real register (not the meta register) for _useVirtAddr
+            # but ultimately return the meta register value
+            rval = e_i386.IntelEmulator.getRegister(self, ridx)
```

Comment:
I'm not entirely sure if we can. We really do need both values. The main value is the "real" value, ie, what we think is really in the register like ax, ebx, ebp, etc. The second value is the full register value (so if the register being retrieved is AH, get the value in eax). We need to always return the first value since that's what the emulator is going to expect. The second value is what we need for tracking taint values, especially through metaregisters.

If we're not dealing with metaregisters, the two values are going to be equal and we only call getRegister once. If we getting a meta register, we need the second call to getRegister to get the full value of the non-meta, 32-bit register. Otherwise due to our taint values and the masking down to 8/16 bits, the taint value would get masked down to 0 and we would miss the usage of it, which then screws up calling convention and other emulator metadata. 

Thankfully the non-meta reg usage is far less common than the first. But you are right in that it is a performance hit. I did some optimizing, but there's still a bit of a slow down. But we do ultimately need both values to satisfy both the low level i386Emulator's real value emulation and the WorkspaceEmulator's register taint tracking. 

But I am open to suggestions on how to make it fast since I am displeased with the slowdown (even though i minized it from what my first attempt this was). But I do like getting all the register usages.

Reply:
good explanation.